### PR TITLE
Include cache metadata on the first line of the repodata cache

### PR DESCRIPTION
**Pull request by chrisburr**
Currently when constructor writes the repodata cache for conda it neglects to add some of the cache headers. Conda is forgiving about this however micromamba is stricter about this and I think it makes sense to make constructor more correct rather than making mamba more lenient.

For reference this is where conda adds these fields to the repodata for `pkgs/cache/*.json`:

https://github.com/conda/conda/blob/05b3e799daba158b74c892bf3aeb60ffe7485e3c/conda/core/subdir_data.py#L822-L834

### CODE REVIEW

File: constructor/conda_interface.py [hunk 1]
```diff
@@ -90,17 +90,28 @@ def get_repodata(url):
...
+        # The first line of the JSON should contain cache metadata
+        # Choose an arbitrary old, expired date, so that conda will want to
+        # immediately update it when not being run in offline mode
+        url = used_repodata.pop('_url').rstrip("/")
+        repodata = json.dumps(used_repodata, indent=2)
+        repodata_header = json.dumps(
+            {
+                "_mod": "Mon, 07 Jan 2019 15:22:15 GMT",
+                "_url": url,
+                "_cache_control": "public, max-age=1200",
```
Comment (wolfv) [inline constructor/conda_interface.py:111]: it turns out that conda also only writes two of these keys (_mod and _url) when fetching any file from a local channel.

I am adding some handling in mamba to make sure that we can also parse that correctly.

Comment (chrisburr) [reply constructor/conda_interface.py:111]: Ah great, I guess that makes sense for local channels. Thanks for dealing with this!

In that case the only thing left for this PR would be to make sure that these keys are on the top line of the file? i.e.

```python
{"_mod": "Mon, 07 Jan 2019 15:22:15 GMT", "_url": url,
  "packaes": {
       ...
  },
  "packaes.conda": {
       ...
  }
}
```

Comment (wolfv) [reply constructor/conda_interface.py:111]: yes, mamba currently expect them at the beginning of the file. Is that not the case here?

Comment (chrisburr) [reply constructor/conda_interface.py:111]: Prior to this PR, constructor just writes `json.dumps(repodata, indent=2)`.

Conda does it in a way that's consistent with `mamba`.

Comment (wolfv) [reply constructor/conda_interface.py:111]: yeah i don't like that ... in that case we'd need to parse the entire file first. Would be nice to correct this in constructor :)

Comment (chrisburr) [reply constructor/conda_interface.py]: I've amended this so that the `_cache_control` and `_etag` fields aren't written.

**Comment by wolfv** [comment]
This looks great to me.

**Review by jezdez — CHANGES_REQUESTED** [review]
This looks good in theory, but needs tests to make sure, we're not regressing in the future on the "strict" behavior that micromamba expects.

**Comment by chrisburr** [comment]
Once this and #489 are merged I'll add `micromamba` tests to the CI.

I've updated miniforge to use this two PRs and it works with the latest micromamba: https://github.com/conda-forge/miniforge/pull/263

**Comment by jezdez** [comment]
@chrisburr With #489 merged this needs another rebase sadly.

**Comment by chrisburr** [comment]
Thanks for taking care of the rebase (and generally picking up the constructor maintenance).

I'll make a PR with tests over the weekend.

**Comment by jezdez** [comment]
@chrisburr No problem, thanks for working on the patches in the first place!

**Pull request merged**